### ISSUE TITLE
Regex Bug

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7786]
Migrated from [rt.perl.org#34064](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34064) (status was 'resolved')

Searchable as RT34064$

**Comment by p5pRT** [comment]
### From mtillberg@comcast.net



&nbsp;   I've found a case where regex failes to properly match a string\. It 
seems to be related to having the start of a match after \(\.\*?\) falling 
on character 32768 \(REG\_INFTY in regcomp\.h plus 1\)\.  If the string is 
has one character added or removed from the beginning\, the regex works 
properly\.  I've found this bug in v5\.8\.6\, v5\.8\.6 compiled with 
\-DDEBUGGING\, and the development v5\.9\.1 with \-DDEBUGGING\.

A simple example is&#8203;:

my $v = sprintf\("%32767sx  x "\, ' '\);
if \($v =~ m/\(\.\*?\)x/\) \{
&nbsp;     print "Match&#8203;: "\.length\($1\)\."\\n";
\}
outputs&#8203;: Match&#8203;: 32770

However\,
my $v = sprintf\("%32766sx  x "\, ' '\);
if \($v =~ m/\(\.\*?\)x/\) \{
&nbsp;     print "Match&#8203;: "\.length\($1\)\."\\n";
\}
outputs&#8203;: Match&#8203;: 32766

When running with \-Dr\, the first case gives&#8203;:

Found floating substr \`x' at offset 32767\.\.\.
\.\.\. snip
&nbsp;   Setting an EVAL scope\, savestack=5
&nbsp;    0 \<> \<            >    |  1&#8203;:  OPEN1
&nbsp;    0 \<> \<            >    |  3&#8203;:  MINMOD
&nbsp;    0 \<> \<            >    |  4&#8203;:  STAR
&nbsp;   Setting an EVAL scope\, savestack=5
&nbsp;                            REG\_ANY can match 32772 times out of 
2147483647\.\.\.
32772 \<       x  x > \<>    |  6&#8203;:    CLOSE1
32772 \<       x  x > \<>    |  8&#8203;:    EXACT \<x>
&nbsp;                               failed\.\.\.
&nbsp;                            REG\_ANY can match 3 times out of 3\.\.\.
32770 \<       x  > \<x >    |  6&#8203;:    CLOSE1
32770 \<       x  > \<x >    |  8&#8203;:    EXACT \<x>
32771 \<       x  x> \< >    | 10&#8203;:    END


The second example gives&#8203;:
Found floating substr \`x' at offset 32766\.\.\.
\.\.\. snip
&nbsp;                            REG\_ANY can match 32766 times out of 32766\.\.\.
32766 \<       > \<x  x >    |  6&#8203;:    CLOSE1
32766 \<       > \<x  x >    |  8&#8203;:    EXACT \<x>
32767 \<       x> \<  x >    | 10&#8203;:    END

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=medium

Site configuration information for perl v5.8.6:

Configured by mt at Sat Feb  5 11:04:58 EST 2005.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
   Platform:
     osname=linux, osvers=2.4.20-8, archname=i686-linux
     uname='linux shuttle 2.4.20-8 #1 thu mar 13 17:54:28 est 2003 i686 
i686 i386 gnulinux '
     config_args=''
     hint=previous, useposix=true, d_sigaction=define
     usethreads=undef use5005threads=undef useithreads=undef 
usemultiplicity=undef
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include 
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm -DDEBUGGING',
     optimize='-O2',
     cppflags='-fno-strict-aliasing -pipe -I/usr/local/include 
-I/usr/include/gdbm -fno-strict-aliasing -pipe -I/usr/local/include 
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm -DDEB
UGGING'
     ccversion='', gccversion='3.2.2 20030222 (Red Hat Linux 3.2.2-5)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib
     libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
     perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
     libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.3.2'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
     cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.6:
     /usr/local/perl-5.8.6/lib/5.8.6/i686-linux
     /usr/local/perl-5.8.6/lib/5.8.6
     /usr/local/perl-5.8.6/lib/site_perl/5.8.6/i686-linux
     /usr/local/perl-5.8.6/lib/site_perl/5.8.6
     /usr/local/perl-5.8.6/lib/site_perl
     .


Environment for perl v5.8.6:
     HOME=/home/mt
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/home/mt/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On 5 Feb 2005 19&#8203;:09&#8203;:38 \-0000\, Mike Tillberg \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > \[\.\.\.\]

You seem to have found a highly significant test case against patch \#11934\.

\-\-\-\-Program\-\-\-\-
my $v = sprintf\("%32767sx  x "\, ' '\);
if \($v =~ m/\(\.\*?\)x/\) \{
&nbsp;     print "Match&#8203;: "\.length\($1\)\."\\n";
\}

\-\-\-\-Output of \.\.\./p2oe1aw/perl\-5\.7\.2@&#8203;11933/bin/perl\-\-\-\-
Match&#8203;: 32767

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pQ9CzJu/perl\-5\.7\.2@&#8203;11934/bin/perl\-\-\-\-
Match&#8203;: 32770

\-\-\-\-EOF \($?='0'\)\-\-\-\-

Change 11934 by jhi@&#8203;alpha on 2001/09/07 12&#8203;:43&#8203;:50

&nbsp;        Subject&#8203;: Re&#8203;: \[PATCH regexec\.c\] more general \.\* and \.\*?
&nbsp;        From&#8203;: "Jeff 'japhy/Marillion' Pinyan" \<jeffp@&#8203;crusoe\.net>
&nbsp;        Date&#8203;: Fri\, 7 Sep 2001 09&#8203;:41&#8203;:10 \-0400 \(EDT\)
&nbsp;        Message\-ID&#8203;: \<Pine\.GSO\.4\.21\.0109070933530\.7041\-100000@&#8203;crusoe\.crusoe\.net>



\-\- 
andreas

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

This seems to be resolved in&#8203;:
\- 5\.8\.9\-to\-be
\- 5\.10\.0
\- blead

So marking it as resolved\.


$ cat rt\-34064\.pl
\#\!/usr/bin/perl

\{
&nbsp;  my $v = sprintf\("%32767sx x "\, ' '\);
&nbsp;  if \($v =~ m/\(\.\*?\)x/\) \{
&nbsp;    print "Match&#8203;: "\.length\($1\)\."\\n";
&nbsp;  \}
\}

\{
&nbsp;  my $v = sprintf\("%32766sx x "\, ' '\);
&nbsp;  if \($v =~ m/\(\.\*?\)x/\) \{
&nbsp;    print "Match&#8203;: "\.length\($1\)\."\\n";
&nbsp;  \}
\}


$ perl\-5\.8\.x/perl rt\-34064\.pl 
Match&#8203;: 32767
Match&#8203;: 32766

$ perl\-5\.10\.x/perl rt\-34064\.pl 
Match&#8203;: 32767
Match&#8203;: 32766

$ perl\-blead/perl rt\-34064\.pl 
Match&#8203;: 32767
Match&#8203;: 32766


$ perl\-5\.8\.8 rt\-34064\.pl
Match&#8203;: 32769
Match&#8203;: 32766

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
